types/node/v4/index.d.ts

```diff
@@ -558,13 +558,17 @@ declare module "http" {
     import * as stream from "stream";
 
     // incoming headers will never contain number
-    export interface IncomingHttpHeaders {
-        [header: string]: string | string[];
+    export type IncomingHttpHeaders = {
+        [header: string]: string
+    } & {
+        'set-cookie'?: string[]
     }
```

Comment:
Could you please split this into two interfaces and and combine them to a type then to allow augmentation. Additionally it would be nice if this could be also added to Node 8 (as the main target here is anyway to sync them),

Reply:
I was on the fence about this honestly. The type names are pretty useless (e.g. we wouldn't expect a docs page on them by themselves) and I've seen IDEs do a worse job at code complete or mouseover docs once the subtypes in a composite type are named.

Reply:
This actually doesn't work. I was adding tests for this and I get the same "Property ''set-cookie'' is incompatible with index signature." I would expect if I added the set cookie property onto IncomingHttpHeaders like this
```
type IncomingHttpHeaders {
    [header: string]: string;
    'set-cookie'?: string[];
}
```
For this to work the index signature must account for string | string[]. Is there any reason to still use an Intersection type?

Reply:
Don't ask me why, but

```
type IncomingHttpHeaders {
    [header: string]: string;
    'set-cookie'?: string[];
}
```

is a compiler error while

```
type IncomingHttpHeaders {
    [header: string]: string;
} & {
    'set-cookie'?: string[];
}
```

is not. Types do not allow properties to have a different value type as the index value type, but you can get around this with composite types. Verified the type definition and variable instantiation compile cleanly on TS 5.3.0.

Reply:
The type definition compiles in the second example, but usage of the type does not. That's why I didn't catch it until I was adding tests for it.

Reply:
What a pity. As we have to include the array in the index signature anyway I see no reason for having two types. To get the best out of it we could add other header names besides `set-cookie` to force them to strings. Users can add more by augmenting the interface for their private header names. e.a.:
```typescript
interface IncomingHttpHeaders extends Partial<Record<"age" | "host" | "date", string>> {
    'set-cookie'?: string[];
    [name: string]: string | string[] | undefined;
}
```
Or using copy/paste instead `Partial<Record<>>` which is maybe easier to read:
```typescript
interface IncomingHttpHeaders1 {
    age?: string;
    host?: string;
    //...
    'set-cookie'?: string[];
    [name: string]: string | string[] | undefined;
}
```
Again the `| undefined` is needed here to allow adding optional properties.

Reply:
Yeah, the list of common headers isn't too long I can add those to string and users can extend.